### PR TITLE
is causal hints for transformer

### PR DESCRIPTION
**Pull request by mikekgfb**
Summary:
make is_causal hint flags available for the top level transformer module.

It's debatable whether this is useful -- at present we autodetect causal masks for src and tgt masks in transformer encoder and decoder, respectively. is_causal flags available woul enable users to short-cut this check by asserting whether they mask is causal, or not.

I am putting this diff up for discussion, not as a solution.  Not doing anything may be the right solution, unless there is strong (data-driven) user demand. -- it appears the consensus is to move ahead with this, as per discussions below.

@cpuhrsch @mikaylagawarecki @jbschlosser @janEbert 

Test Plan: sandcastle

Differential Revision: D47373260

### CODE REVIEW

File: test/test_nn.py [hunk 1]
```diff
@@ -3831,7 +3831,11 @@ def test(encoder_input_shape, decoder_input_shape,
                  src_mask_len=None, tgt_mask_len=None, memory_mask_size=None,
                  src_key_padding_mask_size=None, tgt_key_padding_mask_size=None,
                  memory_key_padding_mask_size=None,
-                 raises=False):
+                 src_is_causal=False, tgt_is_causal=False,
+                 memory_is_causal=False, raises=True):
+
+            raises=True
```
Comment (mikaylagawarecki) [inline test/test_nn.py:3837]: Looks like this was left here by accident

**Review by mikaylagawarecki — COMMENTED** [review]
Please fix lint

File: test/test_nn.py (hunk 1 above)
Comment (mikekgfb) [reply test/test_nn.py:3837]: Actually - this is a trial balloon on removing raises as an argument and just always make it True. ;).

**Comment by ezyang** [comment]
cc'ing @lezcano as we had similar API design questions for things like "is this matrix singular". IIRC our resolution was that adding these kinds of hints made sense and we should be willing to do it.

**Comment by lezcano** [comment]
I don't remember that the specific conversation about "is the matrix singular", but I remember that sort of conversation appeared in two contexts:

[`linalg.lstsq`](https://pytorch.org/docs/stable/generated/torch.linalg.lstsq.html) has a `driver` arg that lets you choose the algorithm that will be used. We wrote in its docs a comprehensive guide on how to choose the correct `driver` arg for your application.

The other discussion we were having was about writing a set of tensor subclasses to mark that certain input matrices have some properties (e.g. they are symmetric or Hermitian, or SPD, or singular...). Then, the function would choose the most efficient algorithm available for a given set of inputs. The user would need to initially "tag" matrices with these classes, and then the algorithms could potentially return tagged matrices.
This approach is nice, because it decouples the properties of the inputs from the API, so it's more extensible. Now, it just makes sense if you expect this API to be extended.

File: torch/nn/modules/transformer.py [hunk 2]
```diff
@@ -132,6 +134,28 @@ def forward(self, src: Tensor, tgt: Tensor, src_mask: Optional[Tensor] = None, t
...
+                compatibility.
+            tgt_is_causal: If specified, applies a causal mask as tgt mask.
+                Default: ``None``; try to detect a causal mask.
+                Warning:
+                ``tgt_is_causal`` provides a hint that ``tgt_mask`` is
+                the causal mask. Providing incorrect hints can result in
+                incorrect execution, including forward and backward
+                compatibility.
+            memory_is_causal: If specified, applies a causal mask as
+                memory mask.
```
[the remark this answers is not in the document]
Comment (janEbert) [reply torch/nn/modules/transformer.py:152]: My opinion was that cross-attention is almost never going to be causal, so it would be a waste to make the check most of the time. In contrast, the encoder self-attention may still be causal when the `TransformerEncoder` is used for a decoder-only model.

**Comment by janEbert** [comment]
See also #98327.

File: torch/nn/modules/transformer.py [hunk 3]
```diff
@@ -863,9 +889,9 @@ def _get_activation_fn(activation: str) -> Callable[[Tensor], Tensor]:
 
 
 def _detect_is_causal_mask(
-        mask: Optional[Tensor],
-        is_causal: Optional[bool],
-        size: Optional[int]
+        mask: Optional[Tensor]=None,
+        is_causal: Optional[bool]=None,
+        size: Optional[int]=None,
```
Comment (janEbert) [inline torch/nn/modules/transformer.py:894]: Not sure how picky you are about code style, but there should be spaces around the equals signs here.

Comment (janEbert) [reply torch/nn/modules/transformer.py:894]: Also, at least the `mask` argument should not be `None` by default, right? Since that one should always be supplied.

**Comment by janEbert** [comment]
I think this change makes sense to have consistency between the Transformer modules. To me, it makes sense that these flags should be available when calling a `Transformer` as well. Otherwise you lose access to these flags if you instantiate a `Transformer`; due to the defaults, users do not need to worry about them (just as before), but can still supply the flags for performance improvements.

File: torch/nn/modules/transformer.py (hunk 3 above)
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:894]: I think it can be because we're not checking whether the mask is None (which is a legit call to forward).  So, then we need to check this condition at all call sites, or allow None.  (Whether it should default to None is a valid ... )

**Review by mikaylagawarecki — COMMENTED** [review]
Agree with the above point that this allows `nn.Transformer` to propagate these kwargs to `TransformerDecoder/TransformerEncoder` and hence `TransformerEncoderLayer/TransformerDecoderLayer` (some of which already had these kwargs in the 2.0 release (e.g. [TransformerDecoderLayer](https://pytorch.org/docs/2.0/generated/torch.nn.TransformerDecoderLayer.html?highlight=transformerdecoderlayer#torch.nn.TransformerDecoderLayer))) so we should add them for consistency

**Comment by mikekgfb** [comment]
> Hm got it, in that case could we just remove it from the function signature and delete the argument at the other places where it is used

That's the plan - I just want to make sure all tests pass, before I hack up lots of call sites, and then have to redo them.  Understood that this crutch can't be upstreamed as-is.  (the only way to run all unit tests is github, since I'm working under fbcode... )

File: test/test_nn.py (hunk 1 above)
Comment (mikekgfb) [reply test/test_nn.py:3837]: > Hm got it, in that case could we just remove it from the function signature and delete the argument at the other places where it is used

That's the plan - I just want to make sure all tests pass, before I hack up lots of call sites, and then have to redo them. Understood that this crutch can't be upstreamed as-is. (the only way to run all unit tests is github, since I'm working under fbcode... )

File: torch/nn/modules/transformer.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikekgfb) [reply torch/nn/modules/transformer.py:152]: Are you arguing it should not even exist as a variable?  (We currently have it here, and for decoder, but only True/False, no auto detection (None) to avoid the overhead of checking, and init to `Flase`, i.e., nobody has to bother.

The question is which one is less surprising, having it (because it's not used frequently) or not having it (for consistency of _mask with _is_causal).

File: test/test_nn.py (hunk 1 above)
Comment (mikaylagawarecki) [reply test/test_nn.py:3837]: gotcha! just flagging :)

**Comment by mikekgfb** [comment]
> Please fix lint

Fixed!

**Review by mikaylagawarecki — APPROVED** [review]
Thanks!

**Comment by mikekgfb** [comment]
@pytorchbot merge


Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
